### DISCUSSION TITLE
Cant get MOOSE installed on Lemhi

### DISCUSSION THREAD

**Discussion by jackeggemeyer** [idaholab/moose#28203] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I am currently following these instructions: https://mooseframework.inl.gov/moose/getting_started/installation/inl_hpc_install_moose.html, and I messed something up. I successfully had MOOSE on Sawtooth. Then I tried to add some custom kernels and make the program, but I accidentally did it when I was in Lemhi, and it broke everything. 

I then tried to "fix" the install by updating it, but I was not in a login node, so I think that did more damage. I thought the best course of action would then be to just "delete" the files and start over, but that, of course, did not fix anything. What should I do to get MOOSE on either Lemhi, sawtooth or both? I attached an image with the error below.

Thank you for your time, and I apologize for the headache!

Also, I'm just curious why the user needs to have separate instances of moose on Lemhi and Sawtooth. Also, I've seen 2 sawtooth hpcs(1 & 2). Do I need a separate instance of MOOSE on sawtooth 1 and 2 as well? 

![image](https://github.com/user-attachments/assets/902e1d88-759a-4dfe-a556-4eea0ce10e3a)

**Comment by GiudGiud** [comment]
Hello

Different moose folder for sawtooth and lemhi for sure.
For sawtooth 1 & 2, I think they are the same.

Which modules are you using?

Guillaume

**Reply by jackeggemeyer** [reply]
Thanks, good to know! I am just going to be using the phase field module and adding some of my coded kernels.

**Reply by GiudGiud** [reply]
which HPC modules? the compilers and mpi but the output of `module list` is what I want to see

**Reply by milljm** [reply]
The reason for the separation is due to, I suspect, the errors you received after the first "whoops". The two machines are different, with different compiler libraries. When you build on Lemhi (`make`), and then accidentally continue to develop and build on Sawtooth (additional `make`s) is when you'll run into compiling issues.

This is why we ask users to create some sort of directory separation (`~/projects/sawtooth`).

As for starting completely over and still hitting a snag... I think we do have issues with all our `moose-dev` modules on _every_ cluster at the moment... Can you try to build on bitterroot within a container instead (we don't have instructions for this yet, so I'll walk you through it)?

Gain a shell on Bitterroot:
<img width="550" alt="Screenshot 2024-07-24 at 6 08 26 AM" src="https://github.com/user-attachments/assets/b59b9ddf-7394-4c05-bd20-cdb47830bb58">

Then perform the following:

```bash
module load use.moose moose-containers moose
moose-shell
```

You'll be placed into a container with all the fixings. Proceed to build your application as normal (you may need to clean your repo before obtaining a successful build).

> [!NOTE]
> The container will have PETSc, libMesh, and WASP already built! So proceed directly to running `make` in your application.

> [!TIP]
> We will eventually include these instructions as the only way to develop with MOOSE. That being said, when we do, there will be no need to create directory separations (`~/projects/sawtooth; ~/projects/lemhi`). Containers are immutable, relocatable objects. They are designed to work everywhere and provide the same experience when doing so.

tagging @loganharbour if he has anything to add

**Comment by jackeggemeyer** [comment]
Thank you so much for the help! @GiudGiud Apologies for the misunderstanding. I was using the moose-dev module, and I am planning on using the Paraview one later.

```
Ran 4392 tests in 596.7 seconds. Average test time 0.6 seconds, maximum test time 22.8 sec
onds.
4392 passed, 99 skipped, 0 pending, 0 failed
moose-x86_64:cf062e7 [eggejohn@bitterroot1: (master) ~/lemhi/projects/moose/test ]
```

I was able to follow the install procedure and it seems to have installed properly! I was wondering now how you submit batch jobs on bitterroot/inl hpcs. My university uses nano and .pbs scripts. An example file would be awesome! Also considering how low on the totem pole of priority I am, do yall have nay recommendations on how many cores I should use to have a good balance of performance and the jobs not waiting in the queue for days? I also don't really know how to execute a moose code without the use of Conda so some clarification on that would also be great. Thank you so much for your help!

And wow, I did all of this accidentally in a folder called Lemhi. Can I change the name of the folder?

I was saving all this in case it didn't work but it did:
The bitterroot sever was already detecting something was up before I even did anything with my moose-dev module as seen below
```
`Lmod has detected the following error:  The following module(s) are unknown:
"cmake/3.29.6-gcc-13.3.0-5udm"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "cmake/3.29.6-gcc-13.3.0-5udm"

Also make sure that all modulefiles written in TCL start with the string #%Module

Executing this command requires loading "cmake/3.29.6-gcc-13.3.0-5udm" which failed while processing the
following module(s):

    Module fullname       Module Filename
    ---------------       ---------------
    moose-dev/2024.06.17  /apps/moose/modules/modulefiles/moose-dev/2024.06.17.lua


Lmod has detected the following error:  The following module(s) are unknown:
"cmake/3.29.6-gcc-13.3.0-5udm"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "cmake/3.29.6-gcc-13.3.0-5udm"

Also make sure that all modulefiles written in TCL start with the string #%Module

Executing this command requires loading "cmake/3.29.6-gcc-13.3.0-5udm" which failed while processing the
following module(s):

    Module fullname       Module Filename
    ---------------       ---------------
    moose-dev/2024.06.17  /apps/moose/modules/modulefiles/moose-dev/2024.06.17.lua


Lmod has detected the following error:  The following module(s) are unknown:
"cmake/3.29.6-gcc-13.3.0-5udm"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "cmake/3.29.6-gcc-13.3.0-5udm"

Also make sure that all modulefiles written in TCL start with the string #%Module

Executing this command requires loading "cmake/3.29.6-gcc-13.3.0-5udm" which failed while processing the
following module(s):

    Module fullname       Module Filename
    ---------------       ---------------
    moose-dev/2024.06.17  /apps/moose/modules/modulefiles/moose-dev/2024.06.17.lua


Lmod has detected the following error:  The following module(s) are unknown:
"cmake/3.29.6-gcc-13.3.0-5udm"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "cmake/3.29.6-gcc-13.3.0-5udm"

Also make sure that all modulefiles written in TCL start with the string #%Module

Executing this command requires loading "cmake/3.29.6-gcc-13.3.0-5udm" which failed while processing the
following module(s):

    Module fullname       Module Filename
    ---------------       ---------------
    moose-dev/2024.06.17  /apps/moose/modules/modulefiles/moose-dev/2024.06.17.lua


Lmod has detected the following error:  The following module(s) are unknown:
"cmake/3.29.6-gcc-13.3.0-5udm"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "cmake/3.29.6-gcc-13.3.0-5udm"

Also make sure that all modulefiles written in TCL start with the string #%Module

Executing this command requires loading "cmake/3.29.6-gcc-13.3.0-5udm" which failed while processing the
following module(s):

    Module fullname       Module Filename
    ---------------       ---------------
    moose-dev/2024.06.17  /apps/moose/modules/modulefiles/moose-dev/2024.06.17.lua


Lmod has detected the following error:  The following module(s) are unknown:
"cmake/3.29.6-gcc-13.3.0-5udm"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "cmake/3.29.6-gcc-13.3.0-5udm"

Also make sure that all modulefiles written in TCL start with the string #%Module

Executing this command requires loading "cmake/3.29.6-gcc-13.3.0-5udm" which failed while processing the
following module(s):

    Module fullname       Module Filename
    ---------------       ---------------
    moose-dev/2024.06.17  /apps/moose/modules/modulefiles/moose-dev/2024.06.17.lua


Lmod has detected the following error:  The following module(s) are unknown:
"cmake/3.29.6-gcc-13.3.0-5udm"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "cmake/3.29.6-gcc-13.3.0-5udm"

Also make sure that all modulefiles written in TCL start with the string #%Module

Executing this command requires loading "cmake/3.29.6-gcc-13.3.0-5udm" which failed while processing the
following module(s):

    Module fullname       Module Filename
    ---------------       ---------------
    moose-dev/2024.06.17  /apps/moose/modules/modulefiles/moose-dev/2024.06.17.lua


Lmod has detected the following error:  The following module(s) are unknown:
"cmake/3.29.6-gcc-13.3.0-5udm"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "cmake/3.29.6-gcc-13.3.0-5udm"

Also make sure that all modulefiles written in TCL start with the string #%Module

Executing this command requires loading "cmake/3.29.6-gcc-13.3.0-5udm" which failed while processing the
following module(s):

    Module fullname       Module Filename
    ---------------       ---------------
    moose-dev/2024.06.17  /apps/moose/modules/modulefiles/moose-dev/2024.06.17.lua


Lmod has detected the following error:  The following module(s) are unknown:
"cmake/3.29.6-gcc-13.3.0-5udm"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "cmake/3.29.6-gcc-13.3.0-5udm"

Also make sure that all modulefiles written in TCL start with the string #%Module

Executing this command requires loading "cmake/3.29.6-gcc-13.3.0-5udm" which failed while processing the
following module(s):

    Module fullname       Module Filename
    ---------------       ---------------
    moose-dev/2024.06.17  /apps/moose/modules/modulefiles/moose-dev/2024.06.17.lua


Lmod has detected the following error:  The following module(s) are unknown:
"cmake/3.29.6-gcc-13.3.0-5udm"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "cmake/3.29.6-gcc-13.3.0-5udm"

Also make sure that all modulefiles written in TCL start with the string #%Module

Executing this command requires loading "cmake/3.29.6-gcc-13.3.0-5udm" which failed while processing the
following module(s):

    Module fullname       Module Filename
    ---------------       ---------------
    moose-dev/2024.06.17  /apps/moose/modules/modulefiles/moose-dev/2024.06.17.lua
```
Then here is me following the instructions to the best of my knowledge, I am not a bash person so I would not be surprised if I messed this up.
```
`[eggejohn@bitterroot1 ~]$ module load use.moose moose-containers moose
[eggejohn@bitterroot1 ~]$ moose-shell
moose-x86_64:cf062e7 [eggejohn@bitterroot1: ~ ]
$ mkdir -p ~/bitteroot/projects
moose-x86_64:cf062e7 [eggejohn@bitterroot1: ~ ]
$ cd lemhi/projects/
moose-x86_64:cf062e7 [eggejohn@bitterroot1: ~/lemhi/projects ]
$ git clone https://github.com/idaholab/moose.git
fatal: destination path 'moose' already exists and is not an empty directory.
moose-x86_64:cf062e7 [eggejohn@bitterroot1: ~/lemhi/projects ]
$ cd moose/
moose-x86_64:cf062e7 [eggejohn@bitterroot1: (master) ~/lemhi/projects/moose ]
$ git chekout master
git: 'chekout' is not a git command. See 'git --help'.

The most similar command is
        checkout
moose-x86_64:cf062e7 [eggejohn@bitterroot1: (master) ~/lemhi/projects/moose ]
$ git checkoutmaster
git: 'checkoutmaster' is not a git command. See 'git --help'.
moose-x86_64:cf062e7 [eggejohn@bitterroot1: (master) ~/lemhi/projects/moose ]
$ checkout master
bash: checkout: command not found
moose-x86_64:cf062e7 [eggejohn@bitterroot1: (master) ~/lemhi/projects/moose ]
$ cd scripts/
moose-x86_64:cf062e7 [eggejohn@bitterroot1: (master) ~/lemhi/projects/moose/scripts ]
$ export MOOSE_JOBS=6 METHODS=opt
moose-x86_64:cf062e7 [eggejohn@bitterroot1: (master) ~/lemhi/projects/moose/scripts ]`
```

I was then able to properly install PETsc but I got these errors when I tried to rebuild libmesh

```
/usr/bin/mkdir -p '/opt/libmesh/include/libmesh'
 /home/eggejohn/lemhi/projects/moose/scripts/../libmesh/build-aux/install-sh -C -m 644 libmesh_config.h ../../include/base/auto_ptr.h ../../include/base/dirichlet_boundaries.h ../../include/base/dof_map.h ../../include/base/dof_object.h ../../include/base/factory.h ../../include/base/float128_shims.h ../../include/base/getpot.h ../../include/base/id_types.h ../../include/base/libmesh.h ../../include/base/libmesh_base.h ../../include/base/libmesh_common.h ../../include/base/libmesh_documentation.h ../../include/base/libmesh_exceptions.h ../../include/base/libmesh_logging.h ../../include/base/libmesh_singleton.h ../../include/base/libmesh_version.h ../../include/base/multi_predicates.h ../../include/base/periodic_boundaries.h ../../include/base/periodic_boundary.h ../../include/base/periodic_boundary_base.h ../../include/base/print_trace.h ../../include/base/reference_counted_object.h ../../include/base/reference_counter.h ../../include/base/single_predicates.h ../../include/base/sparsity_pattern.h ../../include/base/variable.h ../../include/base/variant_filter_iterator.h ../../include/enums/enum_convergence_flags.h ../../include/enums/enum_eigen_solver_type.h ../../include/enums/enum_elem_quality.h ../../include/enums/enum_elem_type.h ../../include/enums/enum_error_estimator_type.h ../../include/enums/enum_fe_family.h ../../include/enums/enum_inf_map_type.h ../../include/enums/enum_io_package.h ../../include/enums/enum_matrix_build_type.h ../../include/enums/enum_norm_type.h ../../include/enums/enum_order.h ../../include/enums/enum_parallel_type.h '/opt/libmesh/include/libmesh'
cp: cannot create regular file '/opt/libmesh/include/libmesh/_inst.898016_': Read-only file system
make[3]: *** [Makefile:1163: install-includeHEADERS] Error 1
make[3]: Leaving directory '/home/eggejohn/lemhi/projects/moose/libmesh/build/include'
make[2]: *** [Makefile:1356: install-am] Error 2
make[2]: Leaving directory '/home/eggejohn/lemhi/projects/moose/libmesh/build/include'
make[1]: *** [Makefile:1190: install-recursive] Error 1
make[1]: Leaving directory '/home/eggejohn/lemhi/projects/moose/libmesh/build/include'
make: *** [Makefile:33743: install-recursive] Error 1
moose-x86_64:cf062e7 [eggejohn@bitterroot1: (master) ~/lemhi/projects/moose/scripts ]
```

**Reply by GiudGiud** [reply]
doesnt bitterrout use slurm? I m not sure you can use pbs

> cp: cannot create regular file '/opt/libmesh/include/libmesh/_inst.898016_': Read-only file system

it's trying to install at the wrong location. I dont know why in /opt.
Did you load a module with libmesh?
Let's unload then let's set LIBMESH_DIR to somewhere reasonable.

**Reply by milljm** [reply]
When you are operating within a container:

> [!NOTE]
> The container will have PETSc, libMesh, and WASP already built! So proceed directly to running `make` in your application.

So do not try and build PETSc, libMesh or WASP yourself. Containers are immutable. You were attempting to modify things _already_ inside the container and thus: permission denied will show up. There is a way to do this, and if you must, we can walk you through it. But I suspect very few folks need this capability.

As I stated previously, there is still something wrong with every `moose-dev` module on every one of our cluster machines (Lemhi, Sawtooth, and Bitterroot). And I am not sure we are going to even attempt to fix it! We've been wanting to move away from native builds on our clusters for some time. Moving forwards, its best to stick with using containers on our HPC machines:

```bash
module load use.moose moose-containers moose
moose-shell
<do application development>
```

As for your question about launching jobs via PBS/Slurm, I'm afraid we only have this working for vetted applications at the moment. I am actively looking into a solution for you, and will report back when I have one. This shouldn't be long... I will make this a priority, as I feel this was an oversight on my part telling you "I will never fix the `moose-dev` module", **AND** not providing a way for you to run developed applications using a job scheduler!

**Reply by jackeggemeyer** [reply]
Thank you for your time and clarification; I am an engineer, not a programmer, so this HPC stuff is very new and confusing to me. Good to know with rebuilding PETSc and the others I would have kept on doing that. 

I am comfortable using some bash to run programs/applications though, so how should I proceed if I want to add one kernel to an application and run some tests on any of the HPCS right now? From the answer, it sounds like I have to wait a little bit or use a container which I need some clarification on because I thought this HPC used modules. Thank you again!

**Reply by milljm** [reply]
Sorry for the delay, I was on vacation last week, allowing myself one hour each day for GitHub discussions/emails... etc.

I think what I need to do is actually fix the current moose-dev modules. I was hoping for a quick solution to using containers for development, but we're just not there yet.

I'll post again when I feel the modules are working.

**Reply by milljm** [reply]
Sawtooth appears to be fine at the moment. Building everything from PETSc to moose/test. Either another from my team fixed the modules, or our HPC team solved an underlying issue.


Can you perform the following (first, gain entry to Sawtooth):
```bash
module purge
module load use.moose moose-dev
```
Next, we're going to clone MOOSE in a completely new area to eliminate a stale build issue:
```bash
mkdir /tmp/my_testing_builds
cd /tmp/my_testing_builds
git clone https://github.com/idaholab/moose
cd moose
git checkout master
```
Next, lets build the stack:
```bash
cd scripts
export METHODS=opt METHOD=opt MOOSE_JOBS=6
./update_and_rebuild_petsc.sh
```
Stop if PETSc failed. If not, proceed:
```bash
./update_and_rebuild_libmesh.sh
```
Stop if libMesh failed. If not, proceed:
```bash
./udpate_and_rebuild_wasp.sh
```
Stop if WASP failed. If not, proceed:
```bash
cd ../test
make -j 6
```
